### PR TITLE
Fix IntelliJ 2021/2024 compatibility issues

### PR DESCRIPTION
**Pull request by pkukielka**
## Changes

Problem: IJ Intellij 2024 `AnAction` classes need to override `getActionUpdateThread` to define if they are running on `EDT` or `BCG` thread:

```kt
override ActionUpdateThread getActionUpdateThread() {
    return ActionUpdateThread.EDT;
  }
```

Unfortunately both `getActionUpdateThread` method in `AnAction` interface, and `ActionUpdateThread` which it returns, were added in IJ `2022.2`, and we still supports IJ `2022.1`.
That means we need to use classes which, for some of the supported by use versions, does not exist.

My solution is based on adding `libs/ActionUpdateThread.jar` library which contains just one simple enum `com.intellij.openapi.actionSystem.ActionUpdateThread`, which is a copy of enum which exsists in IJ versions > `2022.2`.
That allows us to compile it and run targeting `2022.1`. 
But it creates problems for `2022.2` and above, as here `ActionUpdateThread` already exists on the classpath, and IJ classloader does not allow for such duplication and will fail on runtime.
To fix that problem we added class `ActionUpdateThreadClassloaderPatcher`. It uses the fact that JVS loads classes when they are first needed, and removes `ActionUpdateThread.jar` from classpath before it is needed for a first time.
A bit of reflection hacking was still needed, because `reset` method in the meantime changed signature slightly so compile time binding was not working as expected.

## Test plan

Full QA is needed after this change as it touches almost all IJ actions

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -72,6 +72,9 @@ intellij {
 }
 
 dependencies {
+  // ActionUpdateThread.jar contains copy of the
+  // com.intellij.openapi.actionSystem.ActionUpdateThread class
+  implementation(files("libs/ActionUpdateThread.jar"))
```
Comment (mkondratek) [inline build.gradle.kts:79]: nice, but hackish, but nice 🚀

File: src/main/kotlin/com/sourcegraph/cody/statusbar/CodyStatusService.kt [hunk 2]
```diff
@@ -12,12 +12,11 @@ import com.sourcegraph.cody.config.CodyAuthenticationManager
 import com.sourcegraph.cody.ignore.IgnoreOracle
 import com.sourcegraph.common.UpgradeToCodyProNotification
 import com.sourcegraph.config.ConfigUtil
-import javax.annotation.concurrent.GuardedBy
 
 @Service
 class CodyStatusService : CodyStatusListener, Disposable {
 
-  @GuardedBy("this") private var status: CodyStatus = CodyStatus.CodyUninit
+  private var status: CodyStatus = CodyStatus.CodyUninit
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/statusbar/CodyStatusService.kt]: is it ok to just remove it? what does `GuardedBy` do?

File: src/main/kotlin/com/sourcegraph/common/ui/DumbAwareBGTAction.kt [hunk 3]
```diff
@@ -1,17 +1,15 @@
...
-import com.sourcegraph.cody.ui.BGTActionSetter
 import javax.swing.Icon
 import org.jetbrains.annotations.NotNull
 
 abstract class DumbAwareBGTAction : DumbAwareAction {
 
-  constructor() : super() {
-    BGTActionSetter.runUpdateOnBackgroundThread(this)
-  }
+  constructor() : super() {}
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/common/ui/DumbAwareBGTAction.kt]: do we need at all?

**Review by mkondratek — COMMENTED** [review]
code changes look good. let me test it locally...

File: src/main/kotlin/com/sourcegraph/cody/context/ui/ContextToolbarButton.kt [hunk 4]
```diff
@@ -11,8 +11,8 @@ open class ContextToolbarButton(
     private val buttonAction: () -> Unit = {}
 ) : DumbAwareActionButton(name, icon) {
 
-  init {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/context/ui/ContextToolbarButton.kt:14]: I started IU 2024.1 and got this:

```
2024-05-27 11:27:48,267 [   5427] SEVERE - #c.i.o.a.i.ActionUpdater - HelpButton#presentation@ToolbarDecorator (com.sourcegraph.cody.context.ui.HelpButton), text='Help'
com.intellij.diagnostic.PluginException: Cannot load class com.intellij.openapi.actionSystem.ActionUpdateThread (
  error: loader constraint violation: loader com.intellij.ide.plugins.cl.PluginClassLoader @7180c414 wants to load class com.intellij.openapi.actionSystem.ActionUpdateThread. A different class with the same name was previously loaded by com.intellij.util.lang.PathClassLoader @4563e9ab. (com.intellij.openapi.actionSystem.ActionUpdateThread is in unnamed module of loader com.intellij.util.lang.PathClassLoader @4563e9ab),
  classLoader=PluginClassLoader(plugin=PluginDescriptor(name=Cody: AI Coding Assistant with Autocomplete & Chat, id=com.sourcegraph.jetbrains, descriptorPath=plugin.xml, path=~\IdeaProjects\jetbrains\build\idea-sandbox\plugins\Sourcegraph, version=6.0-localbuild, package=null, isBundled=false), packagePrefix=null, state=active)
)
	at com.intellij.ide.plugins.cl.PluginClassLoader.loadClassInsideSelf(PluginClassLoader.kt:331)
	at com.intellij.ide.plugins.cl.PluginClassLoader.tryLoadingClass(PluginClassLoader.kt:178)
	at com.intellij.ide.plugins.cl.PluginClassLoader.loadClass(PluginClassLoader.kt:151)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3402)
	at java.base/java.lang.Class.getMethodsRecursive(Class.java:3543)
	at java.base/java.lang.Class.getMethodsRecursive(Class.java:3557)
	at java.base/java.lang.Class.getMethod0(Class.java:3529)
	at java.base/java.lang.Class.getMethod(Class.java:2225)
	at com.intellij.util.ReflectionUtil.getMethodDeclaringClass(ReflectionUtil.java:203)
	at com.intellij.openapi.actionSystem.ActionClassMetaData.getFlags(ActionClassMetaData.java:55)
	at com.intellij.openapi.actionSystem.ActionClassMetaData.isDefaultUpdate(ActionClassMetaData.java:27)
	at com.intellij.openapi.actionSystem.impl.ActionUpdaterKt.isDefaultImplementationRecursively(ActionUpdater.kt:796)
	at com.intellij.openapi.actionSystem.impl.ActionUpdaterKt.access$isDefaultImplementationRecursively(ActionUpdater.kt:1)
	at com.intellij.openapi.actionSystem.impl.ActionUpdater$updateAction$success$1$1.invokeSuspend(ActionUpdater.kt:560)
	at com.intellij.openapi.actionSystem.impl.ActionUpdater$updateAction$success$1$1.invoke(ActionUpdater.kt)
	at com.intellij.openapi.actionSystem.impl.ActionUpdater$updateAction$success$1$1.invoke(ActionUpdater.kt)
	at com.intellij.openapi.actionSystem.impl.ActionUpdater$updateAction$$inlined$retryOnAwaitSharedData$1.invokeSuspend(ActionUpdater.kt:862)
	at com.intellij.openapi.actionSystem.impl.ActionUpdater$updateAction$$inlined$retryOnAwaitSharedData$1.invoke(ActionUpdater.kt)
	at com.intellij.openapi.actionSystem.impl.ActionUpdater$updateAction$$inlined$retryOnAwaitSharedData$1.invoke(ActionUpdater.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:78)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:167)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source)
	at com.intellij.openapi.actionSystem.impl.ActionUpdater.updateAction(ActionUpdater.kt:888)
	at com.intellij.openapi.actionSystem.impl.ActionUpdater.expandGroupChild(ActionUpdater.kt:412)
	at com.intellij.openapi.actionSystem.impl.ActionUpdater.access$expandGroupChild(ActionUpdater.kt:80)
	at com.intellij.openapi.actionSystem.impl.ActionUpdater$doExpandActionGroup$2$expandResult$1$1$1.invokeSuspend(ActionUpdater.kt:339)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
Caused by: java.lang.LinkageError: loader constraint violation: loader com.intellij.ide.plugins.cl.PluginClassLoader @7180c414 wants to load class com.intellij.openapi.actionSystem.ActionUpdateThread. A different class with the same name was previously loaded by com.intellij.util.lang.PathClassLoader @4563e9ab. (com.intellij.openapi.actionSystem.ActionUpdateThread is in unnamed module of loader com.intellij.util.lang.PathClassLoader @4563e9ab)
	at java.base/java.lang.ClassLoader.defineClass2(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1108)
	at com.intellij.util.lang.UrlClassLoader.consumeClassData(UrlClassLoader.java:291)
	at com.intellij.util.lang.ZipResourceFile.findClass(ZipResourceFile.java:116)
	at com.intellij.util.lang.JarLoader.findClass(JarLoader.java:58)
	at com.intellij.util.lang.ClassPath.findClassInLoader(ClassPath.java:240)
	at com.intellij.util.lang.ClassPath.findClass(ClassPath.java:190)
	at com.intellij.ide.plugins.cl.PluginClassLoader.loadClassInsideSelf(PluginClassLoader.kt:326)
	... 35 more
```

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/context/ui/ContextToolbarButton.kt:14]: Note I needed to change the code in order to get it compiling & running with 2024.1 and IU (in a dev run). Maybe if the plugin is built and installed on that spec, and run in a regular way there is no problem.

**Review by mkondratek — APPROVED** [review]

File: src/main/kotlin/com/sourcegraph/common/ui/DumbAwareBGTAction.kt (hunk 3 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/common/ui/DumbAwareBGTAction.kt]: Unfortunately we do :)

File: src/main/kotlin/com/sourcegraph/cody/context/ui/ContextToolbarButton.kt (hunk 4 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/context/ui/ContextToolbarButton.kt:14]: false alarm - sorry. I changed gradle.properties to switch the version and the type. As I mentioned - built and installed works correctly 🚀 lgtm

File: src/main/kotlin/com/sourcegraph/cody/statusbar/CodyStatusService.kt (hunk 2 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/statusbar/CodyStatusService.kt]: TBH I'm not sure if it was doing anything at this point.
I replaced it with `@Volatile` which should be more correct anyway.

**Pull request merged**